src/ILVerify/src/ILImporter.StackValue.cs

```diff
@@ -165,194 +174,192 @@ public override string ToString()
+            var dstVerType = GetVerificationType(dst);
 
-    if (child.IsType(TI_REF) && parent.IsType(TI_REF)
-        && jitInfo->isSDArray(child.GetClassHandleForObjRef())
-        && jitInfo->isSDArray(parent.GetClassHandleForObjRef()))
-    {
-        BOOL runtime_OK;
+            if (IsAssignable(srcVerType, dstVerType))
+                return true;
 
-        // never be more lenient than the runtime
-        runtime_OK = jitInfo->canCast(child.m_cls, parent.m_cls);
-        if (!runtime_OK)
+            // TODO: Handle other cases: controlled-mutability pointer types, boxed types
             return false;
-
-        CORINFO_CLASS_HANDLE handle;
-        CorInfoType pType = jitInfo->getChildType(child.GetClassHandleForObjRef(),  &handle);
-        CorInfoType cType = jitInfo->getChildType(parent.GetClassHandleForObjRef(), &handle);
-
-        // don't care whether it is signed
-        if (cType == CORINFO_TYPE_NATIVEUINT)
-            cType = CORINFO_TYPE_NATIVEINT;
-        if (pType == CORINFO_TYPE_NATIVEUINT)
-            pType = CORINFO_TYPE_NATIVEINT;
-        
-        if (cType == CORINFO_TYPE_NATIVEINT)
-            return pType == CORINFO_TYPE_NATIVEINT;
-            
-        if (pType == CORINFO_TYPE_NATIVEINT)
-            return cType == CORINFO_TYPE_NATIVEINT;
-
-        return runtime_OK;
-    }
-   
-    if (parent.IsUnboxedGenericTypeVar() || child.IsUnboxedGenericTypeVar())
-    {
-        return (FALSE);  // need to have had child == parent
-    }
-    else if (parent.IsType(TI_REF))
-    {
-        // An uninitialized objRef is not compatible to initialized.
-        if (child.IsUninitialisedObjRef() && !parent.IsUninitialisedObjRef())
-            return FALSE;
-
-        if (child.IsNullObjRef())                   // NULL can be any reference type
-            return TRUE;
-        if (!child.IsType(TI_REF))
-            return FALSE;
-
-        return jitInfo->canCast(child.m_cls, parent.m_cls);
-
-    }
-    else if (parent.IsType(TI_METHOD))
-    {
-        if (!child.IsType(TI_METHOD))
-            return FALSE;
-
-        // Right now we don't bother merging method handles
-        return FALSE;
-    }
-    else if (parent.IsType(TI_STRUCT))
-    {
-        if (!child.IsType(TI_STRUCT))
-            return FALSE;
-
-        // Structures are compatible if they are equivalent
-        return jitInfo->areTypesEquivalent(child.m_cls, parent.m_cls);
-    }
-    else if (parent.IsByRef())
-    {
-        return tiCompatibleWithByRef(jitInfo, child, parent);
-    }
-
-    return FALSE;
-#endif 
         }
 
-
         bool IsBinaryComparable(StackValue src, StackValue dst, ILOpcode op)
         {
-            if (src.Kind == dst.Kind && src.Type == dst.Type)
-                return true;
+            if (src.Kind == dst.Kind)
+            {
+                if (src.Type == dst.Type)
+                    return true;
+
+                if ((src.Type != null && dst.Type != null) &&
```

Comment:
Is this necessary now that the Type is normalized?

Reply:
The `Type` of a StackValue can still be null for null-reference types and in case of 
 `StackValueKind.Unknown`.

Reply:
For reference types, this should be actually doing something else. This method should be implementing logic in `Table III.4: Binary Comparison or Branch Operations`. E.g. `ble` on two identical object reference types should be non-verifiable but it is not caught here today.